File: src/main/services/bedrock-client.ts
import AnthropicBedrock from '@anthropic-ai/bedrock-sdk';
import { STSClient, AssumeRoleCommand } from '@aws-sdk/client-sts';
import { getDecryptedAwsCredentials } from '../ipc/claude';
import { callMcpTool, discoverMcpTools } from './mcp-schema';
import store from '../store';

const DEFAULT_MODEL = 'us.anthropic.claude-sonnet-4-20250514-v1:0';

/** Cached MCP tools in Anthropic tool format */
let cachedTools: Array<{ name: string; description: string; input_schema: Record<string, unknown> }> | null = null;

export async function getBedrockClient(): Promise<{ client: AnthropicBedrock; modelId: string }> {
  const { accessKeyId, secretKey } = getDecryptedAwsCredentials();
  const region: string = (store as any).get('claude.region') || 'us-west-2';
  const roleArn: string = (store as any).get('claude.roleArn') || '';
  const modelId: string = (store as any).get('claude.modelId') || DEFAULT_MODEL;

  let finalAccessKey = accessKeyId;
  let finalSecretKey = secretKey;
  let sessionToken: string | undefined;

  if (roleArn) {
    const sts = new STSClient({
      region,
      credentials: {
        accessKeyId,
        secretAccessKey: secretKey,
      },
    });

    const assumed = await sts.send(
      new AssumeRoleCommand({
        RoleArn: roleArn,
        RoleSessionName: 'teradata-dba-agent-chat',
        DurationSeconds: 3600,
      })
    );

    if (!assumed.Credentials) {
      throw new Error('AssumeRole succeeded but returned no credentials');
    }

    finalAccessKey = assumed.Credentials.AccessKeyId!;
    finalSecretKey = assumed.Credentials.SecretAccessKey!;
    sessionToken = assumed.Credentials.SessionToken;
  }

  const client = new AnthropicBedrock({
    awsRegion: region,
    awsAccessKey: finalAccessKey,
    awsSecretKey: finalSecretKey,
    ...(sessionToken ? { awsSessionToken: sessionToken } : {}),
  });

  return { client, modelId };
}

/** Fetch MCP tools and convert to Anthropic tool format. Caches after first call. */
export async function getMcpToolsForClaude(): Promise<
  Array<{ name: string; description: string; input_schema: Record<string, unknown> }>
> {
  if (cachedTools) return cachedTools;

  try {
    const mcpTools = await discoverMcpTools();
    cachedTools = mcpTools.map((tool) => ({
      name: tool.name,
      description: tool.description,
      input_schema: tool.inputSchema,
    }));
    return cachedTools;
  } catch {
    // MCP not available — return empty, chat works without tools
    return [];
  }
}

/** Execute an MCP tool call by forwarding to the MCP server */
export async function executeTool(name: string, input: Record<string, unknown>): Promise<string> {
  try {
    // Convert all input values to strings for callMcpTool
    const stringArgs: Record<string, string> = {};
    for (const [key, value] of Object.entries(input)) {
      if (value !== null && value !== undefined) {
        stringArgs[key] = typeof value === 'string' ? value : JSON.stringify(value);
      }
    }
    return await callMcpTool(name, stringArgs);
  } catch (err) {
    return `Error: ${err instanceof Error ? err.message : String(err)}`;
  }
}

/** Clear the cached MCP tools — call when MCP config changes */
export function clearToolsCache(): void {
  cachedTools = null;
}

File: evals/harness/bedrock_client.py
"""Bedrock client wrapper for eval harness.

Mirrors the auth flow in src/main/services/bedrock-client.ts.
Reads credentials from environment variables.
"""

from __future__ import annotations

import os

import anthropic


def get_bedrock_client() -> tuple[anthropic.AnthropicBedrock, str]:
    """Create an AnthropicBedrock client using environment credentials.

    Returns:
        Tuple of (client, model_id).

    Environment variables:
        AWS_ACCESS_KEY_ID: Required.
        AWS_SECRET_ACCESS_KEY: Required.
        AWS_SESSION_TOKEN: Optional (for assumed roles).
        AWS_REGION: Default us-west-2.
        AWS_ROLE_ARN: Optional — if set, assumes role via STS first.
        EVAL_MODEL: Model ID override. Default matches app's DEFAULT_MODEL.
    """
    region = os.environ.get("AWS_REGION", "us-west-2")
    model_id = os.environ.get(
        "EVAL_MODEL", "us.anthropic.claude-sonnet-4-20250514-v1:0"
    )
    role_arn = os.environ.get("AWS_ROLE_ARN", "")

    access_key = os.environ.get("AWS_ACCESS_KEY_ID", "")
    secret_key = os.environ.get("AWS_SECRET_ACCESS_KEY", "")
    session_token = os.environ.get("AWS_SESSION_TOKEN")

    if role_arn:
        import boto3

        sts = boto3.client(
            "sts",
            region_name=region,
            aws_access_key_id=access_key,
            aws_secret_access_key=secret_key,
        )
        assumed = sts.assume_role(
            RoleArn=role_arn,
            RoleSessionName="teradata-dba-eval",
            DurationSeconds=3600,
        )
        creds = assumed["Credentials"]
        access_key = creds["AccessKeyId"]
        secret_key = creds["SecretAccessKey"]
        session_token = creds["SessionToken"]

    kwargs: dict = {
        "aws_region": region,
        "aws_access_key": access_key,
        "aws_secret_key": secret_key,
    }
    if session_token:
        kwargs["aws_session_token"] = session_token

    client = anthropic.AnthropicBedrock(**kwargs)
    return client, model_id
